### PR TITLE
Introduce the ability to read runtime configuration from Consul

### PR DESCRIPTION
**Pull request by geoand**

### CODE REVIEW

File: core/deployment/src/main/java/io/quarkus/deployment/builditem/FeatureBuildItem.java [hunk 1]
```diff
@@ -18,6 +18,7 @@
     public static final String CACHE = "cache";
     public static final String CDI = "cdi";
     public static final String CONFIG_YAML = "config-yaml";
+    public static final String CONSUL_CONFIG = "config-config";
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/builditem/FeatureBuildItem.java:21]:
```suggestion
    public static final String CONSUL_CONFIG = "consul-config";
```

File: extensions/consul-config/runtime/pom.xml [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+    <parent>
+        <groupId>io.quarkus</groupId>
+        <artifactId>quarkus-consul-config-parent</artifactId>
+        <version>999-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>quarkus-consul-config</artifactId>
+    <name>Quarkus - Consul Config - Runtime</name>
+    <description>Read runtime configuration from Consul Key - Value stor</description>
```
Comment (gsmet) [inline extensions/consul-config/runtime/pom.xml:15]:
```suggestion
    <description>Read runtime configuration from Consul Key - Value store</description>
```

File: extensions/consul-config/runtime/src/main/java/io/quarkus/consul/config/runtime/ConsulConfig.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    /**
+     * Keys whose value is a raw string.
+     * When this is used, the keys that end up in the user configuration are the keys specified her with '/' replaced by '.'
+     */
+    @ConfigItem
+    Optional<List<String>> rawValueKeys;
+
+    /**
+     * Keys whose value represents a properties file.
+     * When this is used, the keys that end up in the user configuration are the keys of the of the properties file,
```
Comment (gsmet) [inline extensions/consul-config/runtime/src/main/java/io/quarkus/consul/config/runtime/ConsulConfig.java:46]:
```suggestion
     * When this is used, the keys that end up in the user configuration are the keys of the properties file,
```

File: extensions/consul-config/runtime/src/main/java/io/quarkus/consul/config/runtime/ConsulConfig.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+         * Consul token to be provided when authentication is enabled
+         */
+        @ConfigItem
+        Optional<String> token;
+
+        /**
+         * When using HTTPS, whether or not to trust self-signed certificatd
+         */
+        @ConfigItem(defaultValue = "false")
+        boolean trustCerts;
```
Comment (gsmet) [inline extensions/consul-config/runtime/src/main/java/io/quarkus/consul/config/runtime/ConsulConfig.java:98]: Don't we need a trustStore configuration? Also might be a good idea to check other extensions for consistency on this.

**Review by gsmet — CHANGES_REQUESTED** [review]
Spotted a few things here and there.

File: extensions/consul-config/runtime/src/main/java/io/quarkus/consul/config/runtime/ConsulConfig.java (hunk 4 above)
Comment (geoand) [reply extensions/consul-config/runtime/src/main/java/io/quarkus/consul/config/runtime/ConsulConfig.java:98]: Yeah, I was planning on looking into it

Comment (gsmet) [reply extensions/consul-config/runtime/src/main/java/io/quarkus/consul/config/runtime/ConsulConfig.java:98]: I wonder if we should have some sort of shared config group for that. That would enforce consistency.

Comment (geoand) [reply extensions/consul-config/runtime/src/main/java/io/quarkus/consul/config/runtime/ConsulConfig.java:98]: That's what Max was saying IIRC in a mailing thread.

I am personally sceptical, but not completely opposed to it.

Comment (geoand) [reply extensions/consul-config/runtime/src/main/java/io/quarkus/consul/config/runtime/ConsulConfig.java:112]: I added support for a KeyStore file

**Comment by geoand** [comment]
This will of course need documententation

**Comment by gsmet** [comment]
Force pushed a rebase to solve a conflict.

The big question is: should this really be in core? There was a discussion about having an organization to host all sorts of additional extensions. Maybe this one could go there at some point?

I will merge as it has some value to have it in 1.5 but I think we should consider going through the list of all extensions and decide what we keep in core and what we externalize.

**Review by gsmet — APPROVED** [review]

**Comment by geoand** [comment]
@gsmet I agree yes. We should definitely have that discussion and this extension would be a prime target to move.

**Pull request merged**